### DISCUSSION TITLE
LiveData Chart

### DISCUSSION THREAD

**Discussion by DSanjulian** [patrykandpatrick/vico#841] [Questions] [answered]
### Question

I want to implement a linechart showing live data. I get values from a bluetooth device (aprox 10Hz) and create a mutableList of Pairs (value, timestamp). I then use this list in the chart component:

`fun LineChartComponent(points: List<Pair<Float, Float>>) {
    val chartEntryModelProducer = remember {ChartEntryModelProducer()}
    val chartEntries = points.map { entryOf(it.second, it.first) }
    chartEntryModelProducer.setEntries(chartEntries)
    Chart(
        chart = lineChart(),
        chartModelProducer = chartEntryModelProducer,
        startAxis = rememberStartAxis(guideline = null),
        bottomAxis = rememberBottomAxis(guideline = null),
        chartScrollSpec = rememberChartScrollSpec(isScrollEnabled = false),
        runInitialAnimation = false
    )`

This is not working. I tried to limit the size of the list to even just 10 elements, but didn't help. What I am doing wrong?

Thank you!

### Vico version(s)

1.15.0

### UI framework(s)

Jetpack Compose

**Comment by Gowsky** [comment]
Hi @DSanjulian, there are two issues in your code:
1. `setEntries` is called on each recomposition. Make sure to call the function only when the data changes e.g. by using `LaunchedEffect` with an appropriate key.
2. `Chart` uses the default `diffAnimationSpec` whose animation duration is too long for your use case. You need to provide a custom spec that is not longer than 100 ms (given the 10Hz update rate), or a `null` one.

Here's a snippet with the required fixes.
```kt
@Composable
fun LineChartComponent(points: List<Pair<Float, Float>>) {
    val chartEntryModelProducer = remember { ChartEntryModelProducer() }
    LaunchedEffect(points) {
        chartEntryModelProducer.setEntries(points.map { entryOf(it.first, it.second) })
    }
    Chart(
        chart = lineChart(),
        chartModelProducer = chartEntryModelProducer,
        startAxis = rememberStartAxis(guideline = null),
        bottomAxis = rememberBottomAxis(guideline = null),
        chartScrollSpec = rememberChartScrollSpec(isScrollEnabled = false),
        runInitialAnimation = false,
        diffAnimationSpec = null,
    )
}
```

**Comment by DSanjulian** [comment]
Hi Patrick, thanks a lot for the answer!
There's something still wrong, though. The .setEntries seems to not be properly working. I added

`    Log.d("didi", "Points passed: ${points.joinToString()}")
Log.i("didi", "${chartEntryModelProducer.getModel()?.entries}")`

And I get:
![image](https://github.com/user-attachments/assets/b65bf5c8-3f19-415d-8b2a-a953567b8c17)

As you see, even if the list of points keeps growing, when I check the entries of the model there's just one.

**Reply by Gowsky** [reply]
All your entries have _x_ = 0, but each _x_ value must be associated with one _y_ value. It looks like the _x_ and _y_ series are flipped, and you’re using the weight for _x_ and the time for _y_. It should be the other way around.

**Reply by DSanjulian** [reply]
Hi Gowsky, you are totally right, we got the points flipped this time. I am afraid that was not the big issue, though. I changed that, and added a Log in LaunchedEffect that brings some light: LaunchedEffect is being run just once:
![Imagen de WhatsApp 2024-08-11 a las 21 18 20_6de16735](https://github.com/user-attachments/assets/eb0dac9e-a34b-4d1e-a5cb-84e1679d5c65)
![Imagen de WhatsApp 2024-08-11 a las 21 17 57_df14c7c6](https://github.com/user-attachments/assets/663ae6a0-1c9d-4326-8de2-468289e8905d)

(Points passed log still shows first the value and second the time, but this is not how I am setting the Entries)
As you can see We got the update Log just the first iteration

**Reply by Gowsky** [reply]
Your `points` `List` appears to be mutable. It's always the same reference, so it doesn't re-execute the `LaunchedEffect`. Make sure that your `LaunchedEffect` key changes for it to update the chart data.

**Reply by DSanjulian** [reply]
Right, it works when I use points.size as key, but then it only manages to setEntries 3 times:
![Imagen de WhatsApp 2024-08-11 a las 22 03 23_029120bc](https://github.com/user-attachments/assets/8ac1aa5d-01ea-4b38-9f1c-1fa1d2ebf841)

**Reply by Gowsky** [reply]
Does it never return `true` again, even after a couple of seconds?

**Reply by DSanjulian** [reply]
No, once false, always false. And everytime I try it stops when the Model has 3 entries, probably not coincidence, but I can't figure out why :S

**Reply by Gowsky** [reply]
Please provide an MRE, so we can see what's causing this issue. Also, I'd suggest checking if you have the same problem with Vico 2.

**Reply by DSanjulian** [reply]
This is the screen displaying the livedata chart. It gets the values (from a bluetooth device) from BTReader (sends a Pair(data: Float, time: Float) with every datapoint from the BT device.
```kotlin

@Composable
fun LiveDataScreen() {
    val context = LocalContext.current
    val btReader = remember { BTReader(context) }
    var weight by remember { mutableFloatStateOf(0.0f) }
    var maxValue by remember { mutableFloatStateOf(0.0f) }
    var time by remember { mutableFloatStateOf(0f) }
    val graphPoints = remember { mutableStateListOf<Pair<Float, Float>>() } //Sin limitar numero de puntos
  
    LaunchedEffect(Unit) {
        Log.i("Didi", "Iniciando BTReader")
        btReader.startScan()
        btReader.lectura.collect { newEntry ->
            weight = newEntry.first
            time = newEntry.second/1000
            if (weight > maxValue) {
                maxValue = newEntry.first
            }
            graphPoints.add(Pair(weight, time))
        }
    }

    DisposableEffect(Unit) {
        onDispose {
            btReader.stopScan()
        }
    }
    Column(
        modifier = Modifier.fillMaxSize(),
        verticalArrangement = Arrangement.Center,
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Text(
                text = "$weight kg",
                fontSize = 48.sp,
                fontWeight = FontWeight.Bold,
                modifier = Modifier.padding(start = 16.dp)
            )
            Text(
                text = "Max: $maxValue kg",
                fontSize = 24.sp,
                modifier = Modifier.padding(end = 16.dp)
            )
        }
        Spacer(modifier = Modifier.height(8.dp))
        Text(
            text = "${time.roundToInt()} s",
            fontSize = 24.sp,
            modifier = Modifier.padding(end = 16.dp)
        )
        Spacer(modifier = Modifier.height(16.dp))
        Button(onClick = { maxValue = 0.0f }) {
            Text(text = "Reset Max")
        }
        Spacer(modifier = Modifier.height(16.dp))
        LineChartComponent(graphPoints)
        Spacer(modifier = Modifier.height(16.dp))
        Button(onClick = { graphPoints.clear()}) {
            Text(text = "Reset")
        }
    }
}


@Composable
fun LineChartComponent(points: List<Pair<Float, Float>>) {
    val chartEntryModelProducer = remember { ChartEntryModelProducer() }

    LaunchedEffect(points.size) {
        val update = chartEntryModelProducer.setEntries(points.map { entryOf(it.second, it.first) })
        Log.i("didi", "Model update $update")
    }

    Log.d("didi", "Points passed: ${points.joinToString()}")
    Log.i("didi", "Model entries: ${chartEntryModelProducer.getModel()?.entries}")
    Chart(
        chart = lineChart(),
        chartModelProducer = chartEntryModelProducer,
        startAxis = rememberStartAxis(guideline = null),
        bottomAxis = rememberBottomAxis(guideline = null),
        chartScrollSpec = rememberChartScrollSpec(isScrollEnabled = false),
        runInitialAnimation = false,
        diffAnimationSpec = null,
    )
}
```

So far I only tried with 1.15.0 version, I fear it will take me a while to go through v2 documentation to get it working, as I never used it. If in the meantime you can provide me the code to do this in v2 I would appreciate and try it faster.

**Reply by Gowsky** [reply]
@DSanjulian we modified the example (since it's not a proper MRE, as it contains unknown references) and we weren't able to reproduce the problem, even after doubling the frequency. Please provide a **self-contained example** that reproduces the issue.

Also:
- It would be better to move `chartEntryModelProducer` to `LiveDataScreen` so that the `Chart` isn't recomposed on data updates. It listens to changes in the producer, so it will still update.
- Using `getModel` here for debugging is a bad idea and might cause unwanted side effects. The `Boolean` returned by `setEntries` whether the update has been accepted. If you need to log the entries, then you should print the result of `points.map { entryOf(it.second, it.first) }`.

**Reply by DSanjulian** [reply]
Hi @Gowsky,
This is the repository with all the files: [Vico1](https://github.com/DSanjulian/BichoMaker/tree/Timestamp/app/src/main )
The problem is that I am getting the data to show in the chart from a bluetooth device, without it you can't test the app. I took the modelproducer out of the chart composable and  removed the getModel debug, but it was giving interesting results, as the log was not the same as the one from `points.map { entryOf(it.second, it.first) }` It's still not working.

I tried implementing it with Vico2 as well, you can find it here: [Vico2](https://github.com/DSanjulian/BichoMaker/tree/Vico2/app/src/main)
But it isn't working either. Here I think the issue might be with the Dispatcher and its delay, as without it it's so slow, but I am not sure.

Thank you!

**Reply by Gowsky** [reply]
We get 404 when trying to access your repositories. They must be private.

>The problem is that I am getting the data to show in the chart from a bluetooth device

Please create a mock data source then.

**Reply by DSanjulian** [reply]
Ups, sorry, now here you have the project with mock data using [vico1](https://github.com/DSanjulian/BichoMaker/tree/MockVico1/app/src/main/java/com/example/bichomaker) and [vico2](https://github.com/DSanjulian/BichoMaker/tree/Mock/app/src/main/java/com/example/bichomaker).
None working.

Thanks

**Reply by Gowsky** [reply, marked as answer]
First, there are some issues with your usage of Vico and `Flow`. The main problem however was caused by a growing list of _x_ axis labels. Your axis settings added a label for every entry. The number of entries was very large and drawing hundreds (or more) labels slowed down the main thread. Ultimately, such settings don't make sense in this case as all the labels would get truncated. Here's an updated version of `LiveData.kt` (for Vico 2) with issues fixed and suggested bottom axis settings. This should perform as expected.

```kt
package com.example.bichomaker

import MockBTReader
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Button
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.patrykandpatrick.vico.compose.cartesian.CartesianChartHost
import com.patrykandpatrick.vico.compose.cartesian.axis.rememberBottomAxis
import com.patrykandpatrick.vico.compose.cartesian.axis.rememberStartAxis
import com.patrykandpatrick.vico.compose.cartesian.fullWidth
import com.patrykandpatrick.vico.compose.cartesian.layer.rememberLineCartesianLayer
import com.patrykandpatrick.vico.compose.cartesian.rememberCartesianChart
import com.patrykandpatrick.vico.compose.cartesian.rememberVicoScrollState
import com.patrykandpatrick.vico.core.cartesian.HorizontalLayout
import com.patrykandpatrick.vico.core.cartesian.axis.HorizontalAxis
import com.patrykandpatrick.vico.core.cartesian.data.AxisValueOverrider
import com.patrykandpatrick.vico.core.cartesian.data.CartesianChartModelProducer
import com.patrykandpatrick.vico.core.cartesian.data.lineSeries
import kotlin.math.roundToInt
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

@Composable
fun LiveDataScreen() {
    val context = LocalContext.current
    val btReader = remember { MockBTReader(context) }
    var weight by remember { mutableFloatStateOf(0f) }
    var maxValue by remember { mutableFloatStateOf(0f) }
    var time by remember { mutableFloatStateOf(0f) }
    val graphPoints = remember {
        mutableStateListOf<Pair<Float, Float>>()
    } // Sin limitar numero de puntos
    val modelProducer = remember { CartesianChartModelProducer() }

    LaunchedEffect(Unit) {
        withContext(Dispatchers.Default) {
            btReader.lectura.collect { newEntry ->
                weight = newEntry.first
                time = newEntry.second
                if (weight > maxValue) {
                    maxValue = newEntry.first
                }
                graphPoints.add(Pair(weight, time))
                modelProducer.runTransaction {
                    lineSeries {
                        series(x = graphPoints.map { it.second }, y = graphPoints.map { it.first })
                    }
                }
            }
        }
    }

    Column(
        modifier = Modifier.fillMaxSize(),
        verticalArrangement = Arrangement.Center,
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
            Text(
                text = "$weight kg",
                fontSize = 48.sp,
                fontWeight = FontWeight.Bold,
                modifier = Modifier.padding(start = 16.dp)
            )
            Text(
                text = "Max: $maxValue kg",
                fontSize = 24.sp,
                modifier = Modifier.padding(end = 16.dp)
            )
        }
        Spacer(modifier = Modifier.height(8.dp))
        Text(
            text = "${time.roundToInt()} s",
            fontSize = 24.sp,
            modifier = Modifier.padding(end = 16.dp)
        )
        Spacer(modifier = Modifier.height(16.dp))
        Button(onClick = { maxValue = 0.0f }) { Text(text = "Reset Max") }
        Spacer(modifier = Modifier.height(16.dp))
        LineChartComponent(modelProducer)
        Spacer(modifier = Modifier.height(16.dp))
        Button(onClick = { graphPoints.clear() }) { Text(text = "Reset") }
    }
}

@Composable
fun LineChartComponent(modelProducer: CartesianChartModelProducer) {
    CartesianChartHost(
        chart =
            rememberCartesianChart(
                rememberLineCartesianLayer(
                    axisValueOverrider = AxisValueOverrider.auto() // .fixed( 0 20 0 50)
                ),
                startAxis = rememberStartAxis(guideline = null),
                bottomAxis =
                    rememberBottomAxis(
                        guideline = null,
                        itemPlacer =
                            remember {
                                HorizontalAxis.ItemPlacer.default(addExtremeLabelPadding = true)
                            },
                    ),
                horizontalLayout = HorizontalLayout.fullWidth(),
            ),
        modelProducer = modelProducer,
        animationSpec = null,
        runInitialAnimation = false,
        scrollState = rememberVicoScrollState(scrollEnabled = false)
    )
}
```

**Reply by DSanjulian** [reply]
It's working now! Thank you so much, you've been a great help!
One last thing: I am not able to make the graph show the axis label values :S 

Thanks again!

**Reply by Gowsky** [reply]
What axis label values? That's how it's supposed to look with the code we've provided.

<img src="https://github.com/user-attachments/assets/153761d4-85b6-4433-93d8-53d19bff3ed3" width=320/>

**Reply by DSanjulian** [reply]
Weird, in my phone Huawei with EMUI 12.0.0 looks like this:
![Imagen de WhatsApp 2024-08-16 a las 13 19 11_6652d4a4](https://github.com/user-attachments/assets/9b3a607c-246f-452f-9104-c15e61db4c56)

**Reply by Gowsky** [reply]
@DSanjulian it looks like your device is using dark mode which your app doesn't support. Vico adjusts its color to dark mode and uses white color for the labels. You can see it in your screenshot. You have to adjust your screen's background color or the labels' color.

**Comment by DSanjulian** [comment]
That was it, thank you very much @Gowsky!

**Reply by Gowsky** [reply]
You're welcome!
